drv/stm32h7-update-server/src/main.rs

```diff
@@ -0,0 +1,342 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+//
+// Functions for writing to flash for updates
+//
+// This driver is intended to carry as little state as possible. Most of the
+// heavy work and decision making should be handled in other tasks.
+#![no_std]
+#![no_main]
+
+use drv_update_api::UpdateError;
+use idol_runtime::{ClientError, Leased, LenLimit, RequestError, R};
+use ringbuf::*;
+use stm32h7::stm32h753 as device;
+use userlib::*;
+
+// Keys constants are defined in RM0433 Rev 7
+// Section 4.9.2
+const FLASH_KEY1: u32 = 0x4567_0123;
+const FLASH_KEY2: u32 = 0xCDEF_89AB;
+
+// Keys constants are defined in RM0433 Rev 7
+// Section 4.9.3
+const FLASH_OPT_KEY1: u32 = 0x0819_2A3B;
+const FLASH_OPT_KEY2: u32 = 0x4C5D_6E7F;
+
+const BANK_ADDR: u32 = 0x08100000;
+const BANK_END: u32 = 0x08200000;
+
+// Writes are indexed by flash words, BANK_ADDR is word 0,
+// BANK_ADDR + FLASH_WORD_BYTES is word 1 etc.
+const BANK_WORD_LIMIT: usize =
+    (BANK_END - BANK_ADDR) as usize / FLASH_WORD_BYTES;
+
+// RM0433 Rev 7 section 4.3.9
+// Flash word is defined as 256 bits
+const FLASH_WORD_BITS: usize = 256;
+
+// Total length of a word in bytes (i.e. our array size)
+const FLASH_WORD_BYTES: usize = FLASH_WORD_BITS / 8;
+
+// number of 32-bit words we will have to write
+const FLASH_CPU_WORDS: usize = FLASH_WORD_BYTES / 4;
+
+// Block is an abstract concept here. It represents the size of data the
+// driver will process at a time.
+const BLOCK_SIZE_BYTES: usize = FLASH_WORD_BYTES * 32;
+
+#[derive(Copy, Clone, PartialEq)]
+enum Trace {
+    EraseStart,
+    EraseEnd,
+    WriteStart,
+    WriteEnd,
+    FinishStart,
+    FinishEnd,
+    WriteBlock(usize),
+    None,
+}
+
+enum UpdateState {
+    NoUpdate,
+    InProgress,
+    Finished,
+}
+
+ringbuf!(Trace, 64, Trace::None);
+
+const FLASH_TIMEOUT: usize = 10000;
```

Comment:
Nit: this is a retry counter, not a timeout, which I discovered when I went looking for what time unit it's expressed in.